src/graph/FetchVerticesExecutor.cpp

```diff
@@ -155,277 +350,293 @@ void FetchVerticesExecutor::fetchVertices() {
+                    return;
                 }
-                auto tagName = std::move(tagFound).value();
-                auto iter = schema->begin();
-                auto index = 0;
-
-                while (iter) {
-                    auto *field = iter->getName();
-                    auto prop = RowReader::getPropByIndex(reader.get(), index);
-                    if (!ok(prop)) {
-                        LOG(ERROR) << "Read props of tag " << tagName << " failed.";
-                        doError(Status::Error("Read props of tag `%s' failed.", tagName.c_str()));
-                        return;
-                    }
-                    Collector::collectWithoutSchema(value(prop), &writer);
-                    auto colName = folly::stringPrintf("%s.%s", tagName.c_str(), field);
-                    resultColNames_.emplace_back(colName);
-                    returnColNames_.emplace_back(colName);
-                    auto fieldType = iter->getType();
-                    outputSchema->appendCol(std::move(colName), std::move(fieldType));
-                    ++index;
-                    ++iter;
+                record.emplace_back(std::move(value).value());
+            }
+            if (outputSchema == nullptr) {
+                outputSchema = std::make_shared<SchemaWriter>();
+                rsWriter = std::make_unique<RowSetWriter>(outputSchema);
+                auto getSchemaStatus = Collector::getSchema(record, colNames_, colTypes_, outputSchema.get());
+                if (!getSchemaStatus.ok()) {
+                    doError(getSchemaStatus);
+                    return;
                 }
             }
-            if (writer.size() > 1 && rsWriter != nullptr) {
-                rsWriter->addRow(writer.encode());
+            auto writer = std::make_unique<RowWriter>(outputSchema);
+            for (auto& value : record) {
+                auto status = Collector::collect(value, writer.get());
+                if (!status.ok()) {
+                    doError(status);
+                    return;
+                }
             }
+            rsWriter->addRow(*writer);
         }
     }
 
     finishExecution(std::move(rsWriter));
 }
+
+void FetchVerticesExecutor::setupResponse(cpp2::ExecutionResponse &resp) {
+    if (resp_ == nullptr) {
+        resp_ = std::make_unique<cpp2::ExecutionResponse>();
+        resp_->set_column_names(std::move(colNames_));
+    }
+    resp = std::move(*resp_);
+}
+
+void FetchVerticesExecutor::finishExecution(std::unique_ptr<RowSetWriter> rsWriter) {
+    auto outputs = std::make_unique<InterimResult>(std::move(colNames_));
+    if (rsWriter != nullptr) {
+        outputs->setInterim(std::move(rsWriter));
+    }
+
+    if (onResult_) {
+        onResult_(std::move(outputs));
+    } else {
+        resp_ = std::make_unique<cpp2::ExecutionResponse>();
+        auto colNames = outputs->getColNames();
+        resp_->set_column_names(std::move(colNames));
+        if (outputs->hasData()) {
```

Comment:
If it is the rightest sentence,  you could avoid the interim result setting,  it will introduce extra encode/decode. 
You could take toThriftResponse inside GoExecutor for reference.

Reply:
Emm... I simply copy the codes from `FetchExecutor::finishExecution`. Maybe more attention will needed to fully understand.